Source/Scene/CesiumIon.js

```diff
@@ -0,0 +1,267 @@
+     * });
+     *
+     * @example
+     * //Load a CZML file with asset ID of 10890
+     * Cesium.CesiumIon.createResource(10890)
+     *   .then(function (resource) {
+     *     viewer.dataSources.add(Cesium.CzmlDataSource.load(resource));
+     *   });
+     *
+     * @example
+     * //Load an ImageryProvider with asset ID of 2347923
+     * Cesium.CesiumIon.createResource(2347923)
+     *   .then(function (resource) {
+     *     viewer.imageryLayers.addProvider(Cesium.createTileMapServiceImageryProvider({url : resource }));
+     *   });
+     */
+    CesiumIon.createResource = function(assetId, options) {
+        Check.defined('assetId', assetId);
+
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+        var serverUrl = defaultValue(options.serverUrl, CesiumIon.defaultServerUrl);
+        var accessToken = defaultValue(options.accessToken, CesiumIon.defaultAccessToken);
+
+        var resourceOptions = {
+            url: serverUrl + '/v1/assets/' + assetId + '/endpoint'
+        };
+
+        if (defined(accessToken)) {
+            resourceOptions.queryParameters = { access_token: accessToken };
+        }
+
+        var endpointResource = new Resource(resourceOptions);
+        return CesiumIon._loadJson(endpointResource)
+            .then(function(endpoint) {
+                return CesiumIonResource.create(endpoint, endpointResource);
+            });
+    };
+
+    /**
+     * Creates an {@link ImageryProvider} representing a Cesium ion imagery asset.
+     * Unlike {@link CesiumIon.createResource}, this function supports external asset functionality.
+     *
+     * @param {Number} assetId The Cesium ion asset id.
+     * @param {Object} [options] An object with the following properties:
+     * @param {String} [options.accessToken=CesiumIon.defaultAccessToken] The access token to use.
+     * @param {String} [options.serverUrl=CesiumIon.defaultServerUrl] The url to the Cesium ion API server.
+     * @returns {Promise<ImageryProvider>} A promise to an imagery provider presenting the requested Cesium ion Asset.
+     *
+     * @example
+     * //Load an ImageryProvider with asset ID of 2347923
+     * Cesium.CesiumIon.createImageryProvider(2347923)
+     *   .then(function (imageryProvider) {
+     *     viewer.imageryLayers.addProvider(imageryProvider);
+     *   });
+     */
+    CesiumIon.createImageryProvider = function(assetId, options) {
+        return CesiumIon.createResource(assetId, options)
+            .then(function(resource) {
+                return resource.createImageryProvider();
+            });
+    };
+
+    /**
+     * A {@link Resource} instance that encapsulates Cesium ion asset
+     * creation and automatic refresh token handling.
+     *
+     * @private
+     */
+    function CesiumIonResource(options, endpoint, endpointResource) {
+        Resource.call(this, options);
+
+        this.ionData = {
+            endpoint: endpoint,
+            _endpointResource: endpointResource,
```

Comment:
Instead of having this object with private properties, why not just put these private properties on `this`?

Reply:
Two reasons:

1. As I mentioned above, this object is shared across all derived resource.
2. Since this class derives from `Resource`, its better to put everything into a single object to avoid potential collisions if Resource gets a property of the same name added to it.